### PR TITLE
State lockdown

### PR DESCRIPTION
**Pull request by hypnoshock**
# What

The `BaseDispatcher` and `BaseState` have concept of ownership. `BaseState` also has an `allowList` that can be added to by the owner.

# Why

The goal was to only allow certain addresses to be able to mutate game state so that plugin/extension authors cannot cheat by circumventing game rules and breaking the game. 

By default the address passed into the `BaseState` constructor is classed as the `owner` and is able to directly mutate the state of the node graph. It's usually expected that `Rule` contracts are able to manipulate state therefore extra addresses can be 'authorised' by the owner of the `State` contract.

# WIP

Classed as WIP as I haven't made any effort to see if my attempt of locking down state manipulation to authorised parties can be easily circumvented. It appears to work as expected and unit tests are passing. I have made changes to our `ds` project to make use of these updates.

related https://github.com/playmint/ds/pull/1237

### CODE REVIEW

**Comment by chrisfarms** [comment]
might be good to have a test that fails on an unallowed `set` as a regression check

File: contracts/src/BaseState.sol [hunk 1]
```diff
@@ -116,7 +125,7 @@ contract BaseState is State {
         return annotations[nodeID][keccak256(bytes(annotationLabel))];
     }
 
-    function setData(bytes24 nodeID, string memory label, bytes32 data) external {
+    function setData(bytes24 nodeID, string memory label, bytes32 data) external allowListOnly {
```
[the remark this answers is not in the document]
Comment (hypnoshock) [reply contracts/src/BaseState.sol:128]: I'll check but setting data on a building should be using the `SET_DATA_ON_BUILDING` action which is implemented by the `BuildingRule`. If we are using the `setData` function directly we probably need to change that to use a Action/rule

[the remark this answers is not in the document]
Comment (hypnoshock) [reply contracts/src/BaseState.sol:128]: In the `ds` repo I can only see it being called by `BuildingRule`, `CraftingRule` and by the `Schema`. I'm not seeing that it's called by anything else. Am I missing something?

[the remark this answers is not in the document]
Comment (chrisfarms) [reply contracts/src/BaseState.sol:128]: no you are right ... I forgot about the SET_DATA_ON_BUILDING action 🤦

File: contracts/src/BaseState.sol [hunk 2]
```diff
@@ -131,26 +140,24 @@ contract BaseState is State {
         return nodeData[nodeID][keccak256(bytes(annotationLabel))];
     }
 
-    // TODO: allowlist only
-    function registerNodeType(bytes4 kindID, string memory kindName, CompoundKeyKind keyKind) external {
+    function registerNodeType(bytes4 kindID, string memory kindName, CompoundKeyKind keyKind) external allowListOnly {
```
[the remark this answers is not in the document]
Comment (hypnoshock) [reply contracts/src/BaseState.sol]: Nice catch, I've set both `registerNodeType` and `registerEdgeType` as ownerOnly. You're right, rules never register new node or edge types and it makes sense they shouldn't be able to. Searching our repo, only `Downstream.sol` registers nodes and edges so nothing has broken by making it `ownerOnly`

**Review by chrisfarms — APPROVED** [review]

**Pull request merged**